File: content/solid_template/scripts/vscripts/override/CDOTA_Buff.ts
/** @noSelfInFile */
interface ModifierEvent {
	attacker?: CDOTA_BaseNPC;
	caster?: CDOTA_BaseNPC;
	unit?: CDOTA_BaseNPC;
	target?: CDOTA_BaseNPC;
	ability?: CDOTABaseAbility;
	inflictor?: CDOTABaseAbility;
	damage: number;
	damage_type: DAMAGE_TYPES;
	damage_category: DamageCategory_t;
	damage_flags: DOTADamageFlag_t;
	original_damage: number;
	ranged_attack: boolean;
	no_attack_cooldown: boolean;
	record: number;
	fail_type: attackfail;
	do_not_consume?: boolean;
	report_max?: boolean;
	eom_flags?: EOM_DAMAGE_FLAGS;
	attack_states?: ATTACK_STATES;
	itemname?: string;
	gain?: number;
	unitname?: string;
	player_id?: PlayerID;
	loots?: string[];
}
declare interface CDOTA_Buff {
	/**
	 * 获取buff来源技能当前等级键值，如获取不到则会返回0
	 * @param sKey 键名
	 * @returns 值
	 */
	GetAbilitySpecialValueFor(sKey: string): number;
	/**
	 * 获取buff来源技能等级键值，如获取不到则会返回0
	 * @param sKey 键名
	 * @param iLevel 等级，-1为取当前等级，1级填值为0
	 * @returns 值
	 */
	GetAbilityLevelSpecialValueFor(sKey: string, iLevel: number): number;

	/** 伪随机 */
	PRD(chance: number, pseudo_random_recording?: string): boolean;
	/** 增加层数 */
	IncrementStackCount_Engine: typeof CDOTA_Buff.IncrementStackCount;
	IncrementStackCount(iStackCount?: number): void;
	/** 减少层数 */
	DecrementStackCount_Engine: typeof CDOTA_Buff.DecrementStackCount;
	DecrementStackCount(iStackCount?: number): void;
	/** 存Think的表，StartThink用到 */
	_ThinkList: Record<string, string>;
	/**
	 * 自定义的计时器，方便开多个，回调OnThink，name不填默认是buff的名字
	 * @serverOnly
	 */
	StartThink(interval: number, name?: string, callback?: (name: string) => void): void;
	/**
	 * 自定义的计时器的回调
	 * @serverOnly
	 */
	OnThink(name: string): void;
}

CDOTA_Buff.GetAbilitySpecialValueFor = function (sKey: string) {
	if (!IsValid(this)) return 0;
	// @ts-ignore
	let ability = (this.ability as CDOTABaseAbility | undefined) ?? this.GetAbility();
	// @ts-ignore
	if (!IsValid(ability)) return this[sKey] ?? 0;
	let level = ability.GetLevel() - 1;
	if (level == -1) return 0;
	let caster: CDOTA_BaseNPC | undefined = ability.GetCaster();
	if (!IsValid(caster)) {
		caster = this.GetCaster();
	}
	if (!IsValid(caster)) return 0;
	// return AbilityValues.GetAbilityLevelSpecialValueFor(ability.GetAbilityName(), sKey, level, caster, undefined, AbilityValues.GetAbilityOverrideAbilityValue(ability.entindex(), sKey));
	return ability.GetSpecialValueFor(sKey);
};
CDOTA_Buff.GetAbilityLevelSpecialValueFor = function (sKey: string, iLevel: number) {
	if (!IsValid(this)) return 0;
	// @ts-ignore
	let ability = (this.ability as CDOTABaseAbility | undefined) ?? this.GetAbility();
	// @ts-ignore
	if (!IsValid(ability)) return this[sKey] ?? 0;
	return ability.GetLevelSpecialValueFor(sKey, iLevel);
};
CDOTA_Buff.PRD = function (chance, pseudo_random_recording) {
	// @ts-ignore
	return PRD(this.GetCaster(), chance, pseudo_random_recording ?? this.GetName());
};
if (CDOTA_Buff.IncrementStackCount_Engine == undefined) {
	CDOTA_Buff.IncrementStackCount_Engine = CDOTA_Buff.IncrementStackCount;
}
CDOTA_Buff.IncrementStackCount = function (iStackCount?: number) {
	if (iStackCount == undefined) {
		this.IncrementStackCount_Engine();
	} else {
		this.SetStackCount(this.GetStackCount() + iStackCount);
	}
};
if (CDOTA_Buff.DecrementStackCount_Engine == undefined) {
	CDOTA_Buff.DecrementStackCount_Engine = CDOTA_Buff.DecrementStackCount;
}
CDOTA_Buff.DecrementStackCount = function (iStackCount?: number) {
	if (iStackCount == undefined) {
		this.DecrementStackCount_Engine();
	} else {
		this.SetStackCount(this.GetStackCount() - iStackCount);
	}
};
CDOTA_Buff.StartThink = function (interval: number, name?: string, callback?: (name: string) => void) {
	if (IsServer()) {
		const timerName = name ?? this.GetName();
		if (this._ThinkList == undefined) {
			this._ThinkList = {};
		}
		// 传入-1停止计时器
		if (interval == -1) {
			if (this._ThinkList[timerName]) {
				Timer.StopTimer(this._ThinkList[timerName]);
				// @ts-ignore
				this._ThinkList[timerName] = undefined;
			}
			return;
		}
		// 停止同名计时器
		if (this._ThinkList[timerName] != undefined) {
			Timer.StopTimer(this._ThinkList[timerName]);
		}
		const index = Timer.StartIntervalThink(this, interval, () => {
			if (callback != undefined) {
				callback(timerName);
			}
			if (this.OnThink != undefined) {
				this.OnThink(timerName);
			}
		});
		this._ThinkList[timerName] = index;
	}
};

File: game/solid_template/scripts/vscripts/override/CDOTA_Buff.lua
local ____lualib = require("lualib_bundle")
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["4"] = 65,["5"] = 66,["6"] = 66,["8"] = 68,["9"] = 70,["10"] = 70,["11"] = 70,["12"] = 70,["14"] = 70,["16"] = 71,["17"] = 72,["18"] = 72,["20"] = 73,["21"] = 74,["22"] = 75,["24"] = 77,["25"] = 77,["27"] = 79,["28"] = 65,["29"] = 81,["30"] = 82,["31"] = 82,["33"] = 84,["34"] = 86,["35"] = 86,["36"] = 86,["37"] = 86,["39"] = 86,["41"] = 87,["42"] = 81,["43"] = 89,["44"] = 91,["45"] = 91,["46"] = 91,["47"] = 91,["48"] = 91,["49"] = 91,["50"] = 89,["51"] = 93,["52"] = 94,["54"] = 96,["55"] = 97,["56"] = 98,["58"] = 100,["60"] = 96,["61"] = 103,["62"] = 104,["64"] = 106,["65"] = 107,["66"] = 108,["68"] = 110,["70"] = 106,["71"] = 113,["72"] = 114,["73"] = 115,["74"] = 116,["75"] = 117,["77"] = 120,["78"] = 121,["79"] = 122,["80"] = 124,["84"] = 129,["85"] = 130,["87"] = 132,["88"] = 132,["89"] = 132,["90"] = 132,["91"] = 133,["92"] = 134,["94"] = 136,["95"] = 137,["97"] = 132,["98"] = 132,["99"] = 140,["101"] = 113});
CDOTA_Buff.GetAbilitySpecialValueFor = function(self, sKey)
    if not IsValid(self) then
        return 0
    end
    local ability = self.ability or self:GetAbility()
    if not IsValid(ability) then
        local ____self_sKey_0 = self[sKey]
        if ____self_sKey_0 == nil then
            ____self_sKey_0 = 0
        end
        return ____self_sKey_0
    end
    local level = ability:GetLevel() - 1
    if level == -1 then
        return 0
    end
    local caster = ability:GetCaster()
    if not IsValid(caster) then
        caster = self:GetCaster()
    end
    if not IsValid(caster) then
        return 0
    end
    return ability:GetSpecialValueFor(sKey)
end
CDOTA_Buff.GetAbilityLevelSpecialValueFor = function(self, sKey, iLevel)
    if not IsValid(self) then
        return 0
    end
    local ability = self.ability or self:GetAbility()
    if not IsValid(ability) then
        local ____self_sKey_1 = self[sKey]
        if ____self_sKey_1 == nil then
            ____self_sKey_1 = 0
        end
        return ____self_sKey_1
    end
    return ability:GetLevelSpecialValueFor(sKey, iLevel)
end
CDOTA_Buff.PRD = function(self, chance, pseudo_random_recording)
    return PRD(
        nil,
        self:GetCaster(),
        chance,
        pseudo_random_recording or self:GetName()
    )
end
if CDOTA_Buff.IncrementStackCount_Engine == nil then
    CDOTA_Buff.IncrementStackCount_Engine = CDOTA_Buff.IncrementStackCount
end
CDOTA_Buff.IncrementStackCount = function(self, iStackCount)
    if iStackCount == nil then
        self:IncrementStackCount_Engine()
    else
        self:SetStackCount(self:GetStackCount() + iStackCount)
    end
end
if CDOTA_Buff.DecrementStackCount_Engine == nil then
    CDOTA_Buff.DecrementStackCount_Engine = CDOTA_Buff.DecrementStackCount
end
CDOTA_Buff.DecrementStackCount = function(self, iStackCount)
    if iStackCount == nil then
        self:DecrementStackCount_Engine()
    else
        self:SetStackCount(self:GetStackCount() - iStackCount)
    end
end
CDOTA_Buff.StartThink = function(self, interval, name, callback)
    if IsServer() then
        local timerName = name or self:GetName()
        if self._ThinkList == nil then
            self._ThinkList = {}
        end
        if interval == -1 then
            if self._ThinkList[timerName] then
                Timer:StopTimer(self._ThinkList[timerName])
                self._ThinkList[timerName] = nil
            end
            return
        end
        if self._ThinkList[timerName] ~= nil then
            Timer:StopTimer(self._ThinkList[timerName])
        end
        local index = Timer:StartIntervalThink(
            self,
            interval,
            function()
                if callback ~= nil then
                    callback(timerName)
                end
                if self.OnThink ~= nil then
                    self:OnThink(timerName)
                end
            end
        )
        self._ThinkList[timerName] = index
    end
end
